core/src/main/java/com/linecorp/armeria/internal/common/SslContextFactory.java

```diff
@@ -0,0 +1,306 @@
+        }
+    }
+
+    private void applyTlsConfig(SslContextBuilder contextBuilder) {
+        if (tlsConfig == null) {
+            return;
+        }
+
+        if (tlsConfig instanceof ServerTlsConfig) {
+            final ServerTlsConfig serverTlsConfig = (ServerTlsConfig) tlsConfig;
+            contextBuilder.clientAuth(serverTlsConfig.clientAuth());
+        } else {
+            final ClientTlsConfig clientTlsConfig = (ClientTlsConfig) tlsConfig;
+            if (clientTlsConfig.tlsNoVerifySet()) {
+                contextBuilder.trustManager(InsecureTrustManagerFactory.INSTANCE);
+            } else if (!clientTlsConfig.insecureHosts().isEmpty()) {
+                contextBuilder.trustManager(IgnoreHostsTrustManager.of(clientTlsConfig.insecureHosts()));
+            }
+        }
+        tlsConfig.tlsCustomizer().accept(contextBuilder);
+    }
+
+    private MeterIdPrefix meterIdPrefix(SslContextMode mode) {
+        MeterIdPrefix meterIdPrefix = this.meterIdPrefix;
+        if (meterIdPrefix == null) {
+            if (mode == SslContextMode.SERVER) {
+                meterIdPrefix = SERVER_METER_ID_PREFIX;
+            } else {
+                meterIdPrefix = CLIENT_METER_ID_PREFIX;
+            }
+        }
+        return meterIdPrefix;
+    }
+
+    @VisibleForTesting
+    public int numCachedContexts() {
+        return cache.size();
+    }
+
+    public enum SslContextMode {
+        SERVER,
+        CLIENT_HTTP1_ONLY,
+        CLIENT
+    }
+
+    private static final class CacheKey {
+        private final SslContextMode mode;
+        @Nullable
+        private final TlsKeyPair tlsKeyPair;
+
+        private CacheKey(SslContextMode mode, @Nullable TlsKeyPair tlsKeyPair) {
+            this.mode = mode;
+            this.tlsKeyPair = tlsKeyPair;
+        }
+
+        SslContextMode mode() {
+            return mode;
+        }
+
+        @Nullable
+        TlsKeyPair tlsKeyPair() {
+            return tlsKeyPair;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (!(o instanceof CacheKey)) {
+                return false;
+            }
+            final CacheKey that = (CacheKey) o;
+            return mode == that.mode && Objects.equals(tlsKeyPair, that.tlsKeyPair);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(mode, tlsKeyPair);
+        }
+
+        @Override
+        public String toString() {
+            return MoreObjects.toStringHelper(this)
+                              .omitNullValues()
+                              .add("mode", mode)
+                              .add("tlsKeyPair", tlsKeyPair)
+                              .toString();
+        }
+    }
+
+    private static final class SslContextHolder {
+        private final SslContext sslContext;
+        @Nullable
+        private final CloseableMeterBinder meterBinder;
```

Comment:
It seems like this `meterBinder` is never bound to a meter registry. 🤔

Reply:
😅

Reply:
Added a test for that.